File: apps/nihongo_flutter/lib/domain/entity/vocabulary.dart
import 'sync_status.dart';

class Vocabulary {
  const Vocabulary({
    required this.id,
    required this.lessonNumber,
    required this.kana,
    this.kanji,
    required this.meaning,
    required this.romaji,
    required this.sortOrder,
    this.syncStatus = SyncStatus.synced,
  });

  final int id;
  final int lessonNumber;
  final String kana;
  final String? kanji;
  final String meaning;
  final String romaji;
  final int sortOrder;
  final SyncStatus syncStatus;

  Vocabulary copyWith({
    SyncStatus? syncStatus,
  }) {
    return Vocabulary(
      id: id,
      lessonNumber: lessonNumber,
      kana: kana,
      kanji: kanji,
      meaning: meaning,
      romaji: romaji,
      sortOrder: sortOrder,
      syncStatus: syncStatus ?? this.syncStatus,
    );
  }
}

File: apps/nihongo-ios/NihongoEDU/Domain/Vocabulary.swift
import SwiftData
import Foundation

@Model
final class Vocabulary {
    @Attribute(.unique) var id: Int
    var lessonNumber: Int
    var kana: String
    var kanji: String?
    var meaning: String
    var romaji: String
    var sortOrder: Int
    var updatedAt: Date

    @Relationship(deleteRule: .cascade, inverse: \SRSCard.vocabulary)
    var srsCard: SRSCard?

    init(
        id: Int,
        lessonNumber: Int,
        kana: String,
        kanji: String? = nil,
        meaning: String,
        romaji: String = "",
        sortOrder: Int = 0
    ) {
        self.id = id
        self.lessonNumber = lessonNumber
        self.kana = kana
        self.kanji = kanji
        self.meaning = meaning
        self.romaji = romaji
        self.sortOrder = sortOrder
        self.updatedAt = .now
    }
}
